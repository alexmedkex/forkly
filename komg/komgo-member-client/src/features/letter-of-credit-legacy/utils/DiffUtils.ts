import { ICargo, ITrade, IDiff } from '@komgo/types'
import { compare } from 'fast-json-patch'
import { ReplaceOperation } from 'fast-json-patch/lib/core'
import { get } from 'lodash'
import { ILetterOfCredit } from '../types/ILetterOfCredit'

const BLACK_LIST = ['createdAt', 'updatedAt', 'deletedAt', '__v', '_id']

const toDiff = (op: ReplaceOperation<any>, previous, type: 'ITrade' | 'ICargo' | 'ILC'): IDiff => {
  const path = op.path.split('/').slice(1)
  return {
    op: op.op,
    path: op.path,
    value: op.value,
    oldValue: get(previous, path),
    type
  }
}

// Remove a list of properties from an object.
export const omitDeep = (obj = {}, keys: string[]) => {
  return keys.reduce((memo: any, key: string) => {
    return omitKeyDeep(memo, key)
  }, obj)
}

function omitKeyDeep(obj = {}, key) {
  if (Array.isArray(obj)) {
    return omitDeepArrayWalk(obj, key)
  }
  const keys = Object.keys(obj)
  const newObj = {}
  keys.forEach(i => {
    if (i !== key) {
      const val = obj[i]
      if (Array.isArray(val)) {
        newObj[i] = omitDeepArrayWalk(val, key)
      } else if (typeof val === 'object' && val !== null) {
        newObj[i] = omitKeyDeep(val, key)
      } else {
        newObj[i] = val
      }
    }
  })
  return newObj
}

function omitDeepArrayWalk(arr, key) {
  return arr.map(val => {
    if (Array.isArray(val)) {
      return omitDeepArrayWalk(val, key)
    } else if (typeof val === 'object') {
      return omitKeyDeep(val, key)
    }
    return val
  })
}

export const tradeDiff = (previous: ITrade, latest: ITrade): IDiff[] => {
  return !previous || !latest
    ? []
    : compare(omitDeep(previous, BLACK_LIST), omitDeep(latest, BLACK_LIST)).map((op: ReplaceOperation<any>) =>
        toDiff(op, previous, 'ITrade')
      )
}

export const cargoDiff = (previous: ICargo, latest: ICargo): IDiff[] => {
  return !previous || !latest
    ? []
    : compare(omitDeep(previous, BLACK_LIST), omitDeep(latest, BLACK_LIST)).map((op: ReplaceOperation<any>) =>
        toDiff(op, previous, 'ICargo')
      )
}

export const letterOfCreditDiff = (previous: ILetterOfCredit, latest: ILetterOfCredit): IDiff[] =>
  compare(omitDeep(previous, BLACK_LIST), omitDeep(latest, BLACK_LIST)).map((op: ReplaceOperation<any>) =>
    toDiff(op, previous, 'ILC')
  )
