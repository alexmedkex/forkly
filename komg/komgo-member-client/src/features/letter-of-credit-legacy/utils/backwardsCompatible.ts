import { ILetterOfCredit } from '../types/ILetterOfCredit'
import { ICargo, ITrade } from '@komgo/types'

export const withTradeAndCargoSnapshotSourceIds = (letterOfCredit?: ILetterOfCredit): ILetterOfCredit =>
  letterOfCredit && {
    ...letterOfCredit,
    tradeAndCargoSnapshot: letterOfCredit.tradeAndCargoSnapshot && {
      ...letterOfCredit.tradeAndCargoSnapshot,
      trade: letterOfCredit.tradeAndCargoSnapshot.trade && withSourceId(letterOfCredit.tradeAndCargoSnapshot.trade),
      cargo: letterOfCredit.tradeAndCargoSnapshot.cargo && withSourceId(letterOfCredit.tradeAndCargoSnapshot.cargo)
    }
  }

interface VaktId {
  vaktId?: string
}

const withSourceId = <T extends ITrade | ICargo>({ vaktId, ...payload }: T & VaktId): T =>
  ({
    ...payload,
    sourceId: (payload as T).sourceId || vaktId
  } as T)
