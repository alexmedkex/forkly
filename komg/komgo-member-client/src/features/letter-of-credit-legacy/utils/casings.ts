import { sentenceCase } from '../../../utils/casings'

export const sentenceCaseWithLC = (str: string = '') => {
  return sentenceCase(str)
    .split(' ')
    .map(word => (word === 'lc' ? 'LC' : word))
    .join(' ')
}
