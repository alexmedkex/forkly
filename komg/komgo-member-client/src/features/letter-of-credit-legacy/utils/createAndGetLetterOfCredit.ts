import axios from '../../../utils/axios'
import { TRADE_FINANCE_BASE_ENDPOINT, DOCUMENTS_BASE_ENDPOINT } from '../../../utils/endpoints'
import { ILetterOfCreditTemplate } from '../constants'
import { AxiosResponse } from 'axios'

export interface CreateLetterOfCreditPayload {
  templateId: string
  fields: ILetterOfCreditTemplate
}

export interface GetLetterOfCreditPayload {
  id: string
}

export const createAndGetLetterOfCreditDocument = (payload: CreateLetterOfCreditPayload): Promise<AxiosResponse> => {
  return axios.post(`${DOCUMENTS_BASE_ENDPOINT}/document-templates/generate-document`, payload, {
    responseType: 'arraybuffer'
  })
}

export const getLetterOfCreditDocument = async (payload: GetLetterOfCreditPayload): Promise<AxiosResponse> => {
  const { data: documents } = await axios.get(`${TRADE_FINANCE_BASE_ENDPOINT}/lc/${payload.id}/documents`)
  if (!documents.length) {
    throw new Error(`No document with ID=${payload.id} was found`)
  }
  return axios.get(`${TRADE_FINANCE_BASE_ENDPOINT}/lc/documents/${documents[0].id}/content`, {
    responseType: 'arraybuffer'
  })
}
