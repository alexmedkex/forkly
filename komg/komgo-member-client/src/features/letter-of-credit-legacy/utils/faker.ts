import {
  IDiff,
  PaymentTermsEventBase,
  PaymentTermsWhen,
  PaymentTermsTimeUnit,
  PaymentTermsDayType,
  Currency,
  PriceUnit,
  InvoiceQuantity,
  CreditRequirements,
  buildFakeCargo,
  Grade,
  PaymentTermsOption
} from '@komgo/types'
import { ILetterOfCredit, ILetterOfCreditStatus } from '../../letter-of-credit-legacy/types/ILetterOfCredit'
import { IStateTransition } from '../../letter-of-credit-legacy/store/types'
import { IMember } from '../../members/store/types'
import { Roles } from '../../letter-of-credit-legacy/constants/roles'
import { ITrade, ICargo } from '@komgo/types'
import { Document, Product } from '../../document-management/store/types'
import {
  findTasksByLetterOfCreditId,
  findTaskStatusByLetterOfCreditId,
  findLatestShipment
} from '../../letter-of-credit-legacy/utils/selectors'
import { findParticipantCommonNames, findRole } from '../../financial-instruments/utils/selectors'
import { ILetterOfCreditEnriched } from '../../letter-of-credit-legacy/containers/LetterOfCreditDashboard'
import { Task, TaskStatus } from '../../tasks/store/types'
import { LetterOfCreditTaskType } from '../../letter-of-credit-legacy/constants/taskType'
import * as immutable from 'immutable'
import { CouneterpartyStatus, Counterparty } from '../../counterparties/store/types'
import {
  TEMPLATE_TYPE_OPTIONS,
  INVOICE_REQUIREMENT_OPTIONS,
  BILL_OF_LADING_ENDORSEMENT_OPTIONS,
  FEES_PAYABLE_BY_OPTIONS,
  AVAILABLE_WITH_OPTIONS
} from '../constants'
import { ITradeAndCargoSnapshot } from '../../letter-of-credit-legacy/types/ITradeAndCargoSnapshot'
import { mapActionsToTaskType } from './selectors'
import { initialLetterOfCreditValues } from '../constants'
import { ModeOfTransport, Law, TradeSource } from '@komgo/types'
import { ILCPresentation } from '../types/ILCPresentation'
import { LCPresentationStatus } from '../store/presentation/types'
import { LOC_CATEGORIES, TradingRole } from '../../trades/constants'
import { ITimer } from '../../../store/common/types'
import { BottomSheetStatus } from '../../bottom-sheet/store/types'
import { ITradeEnriched } from '../../trades/store/types'

const APPLICANT_ID = 'cf63c1f8-1165-4c94-a8f8-9252eb4f0016'
const BENEFICIARY_ID = '08e9f8e3-94e5-459e-8458-ab512bee6e2c'
const SOME_DATE = '2019-02-04T14:47:26.988Z'

export const fakeLetterOfCredit = ({
  _id = '08e9f8',
  transactionHash = '0x001',
  reference = 'LC18-MER-1',
  issuingBankReference = 'BK18-XX-1',
  applicantId = APPLICANT_ID,
  beneficiaryId = BENEFICIARY_ID,
  issuingBankId = '1bc05a66-1eba-44f7-8f85-38204e4d3516',
  beneficiaryBankId = 'ecc3b179-00bc-499c-a2f9-f8d1cc58e9db',
  amount = 1000000,
  status = ILetterOfCreditStatus.REQUESTED,
  beneficiaryBankRole = Roles.ADVISING_BANK,
  tradeAndCargoSnapshot = fakeTradeAndCargoSnapshot(),
  updatedAt = '2018-11-13T00:00:00.000Z'
} = {}): ILetterOfCredit => ({
  _id,
  transactionHash,
  applicantId,
  applicantContactPerson: 'Donald Duck',
  beneficiaryId,
  beneficiaryContactPerson: 'string',
  issuingBankId,
  issuingBankContactPerson: 'Scrooge',
  issuingBankReference,
  direct: false,
  beneficiaryBankId,
  // advisingBankId,
  // negotiatingBankId,
  beneficiaryBankContactPerson: 'Mickey Mouse',
  beneficiaryBankRole,
  tradeId: '5bc70b98a4d2e71359efd4ab',
  tradeAndCargoSnapshot,
  type: 'IRREVOCABLE',
  applicableRules: 'UCP latest version',
  feesPayableBy: FEES_PAYABLE_BY_OPTIONS.APPLICANT,
  currency: 'EUR',
  amount,
  expiryDate: '2020-12-31',
  expiryPlace: Roles.ISSUING_BANK,
  availableWith: AVAILABLE_WITH_OPTIONS.ISSUING_BANK,
  availableBy: 'DEFERRED_PAYMENT',
  partialShipmentAllowed: true,
  transhipmentAllowed: false,
  documentPresentationDeadlineDays: 21,
  comments: 'a comment',
  reference,
  status,
  freeTextLc: 'a template',
  billOfLadingEndorsement: BILL_OF_LADING_ENDORSEMENT_OPTIONS.ISSUING_BANK,
  invoiceRequirement: INVOICE_REQUIREMENT_OPTIONS.EXHAUSTIVE,
  templateType: TEMPLATE_TYPE_OPTIONS.FREE_TEXT,

  LOI: 'template',
  updatedAt
})

export const fakeLetterOfCreditTemplate = (fields: object = {}) => ({
  ...initialLetterOfCreditValues,
  applicant: { name: '', locality: '' },
  beneficiary: { name: '', locality: '' },
  beneficiaryBank: { name: '', locality: '' },
  issuingBank: { name: '', locality: '' },
  cargo: {},
  cargoIds: [],
  direct: false,
  LOIAllowed: false,
  LOI: '',
  applicableRules: '',
  availableBy: '',
  ...fields
})

export const fakeMember = ({
  staticId = APPLICANT_ID,
  commonName = 'Applicant Name',
  isMember = true,
  isFinancialInstitution = false,
  country = 'country'
} = {}): IMember => {
  return {
    staticId,
    isMember,
    isFinancialInstitution,
    vaktStaticId: 'V234',
    hasSWIFTKey: false,
    x500Name: {
      CN: commonName,
      O: `${commonName} ltd`,
      C: country,
      L: 'city',
      STREET: 'street',
      PC: 'postal code'
    }
  }
}

export const fakeCounterparty = ({
  staticId = APPLICANT_ID,
  commonName = 'Applicant Name',
  isMember = true,
  isFinancialInstitution = false,
  status = 'Onboarded',
  country = 'country'
} = {}): Counterparty => {
  return {
    covered: status === CouneterpartyStatus.COMPLETED ? true : false,
    status,
    ...fakeMember({ staticId, commonName, isMember, isFinancialInstitution, country })
  }
}

export const fakeTrade = ({
  _id = '123',
  source = TradeSource.Vakt,
  sourceId = 'E1243',
  buyer = 'BP-id',
  buyerEtrmId = 'E222',
  seller = 'Mercuria-d',
  sellerEtrmId = '',
  price = 73.1415,
  creditRequirement = CreditRequirements.DocumentaryLetterOfCredit,
  deliveryLocation = 'Weymouth',
  tradingRole = TradingRole.BUYER
} = {}): ITradeEnriched => ({
  _id,
  source,
  sourceId,
  status: LOC_CATEGORIES.TO_BE_FINANCED,
  buyer,
  seller,
  price,
  deliveryPeriod: {
    startDate: '2020-12-1',
    endDate: '2020-12-31'
  },
  buyerEtrmId,
  sellerEtrmId,
  dealDate: '2018-10-29',
  paymentTerms: {
    eventBase: 'BL',
    when: PaymentTermsWhen.From,
    time: 30,
    timeUnit: PaymentTermsTimeUnit.Days,
    dayType: PaymentTermsDayType.Calendar
  },
  currency: Currency.USD,
  priceUnit: PriceUnit.BBL,
  quantity: 600000,
  deliveryTerms: 'FOB',
  minTolerance: 1,
  maxTolerance: 1,
  invoiceQuantity: InvoiceQuantity.Load,
  generalTermsAndConditions: 'SUKO90',
  laytime: 'as per GT&Cs',
  demurrageTerms: 'as per GT&Cs',
  law: Law.EnglishLaw,
  commodity: 'BFOET',
  creditRequirement,
  requiredDocuments: [
    'BILL_OF_LADING',
    'CERTIFICATE_OF_ORIGIN',
    'QUALITY_AND_QUANTITY_REPORT',
    'CERTIFICATE_OF_INSURANCE'
  ],
  deliveryLocation,
  createdAt: '',
  updatedAt: '',
  tradingRole
})

export const fakeTradeSeller = ({
  _id = '123',
  source = TradeSource.Vakt,
  sourceId = 'E1243',
  seller = 'BP-id',
  sellerEtrmId = 'E111',
  buyer = 'Mercuria-d',
  price = 73.1415,
  deliveryLocation = 'Weymouth',
  tradingRole = TradingRole.SELLER
} = {}): ITradeEnriched => ({
  _id,
  source,
  sourceId,
  status: 'OK',
  buyer,
  seller,
  price,
  deliveryPeriod: {
    startDate: '2019-01-02',
    endDate: '2019-01-05'
  },
  buyerEtrmId: '',
  sellerEtrmId,
  dealDate: '2018-10-29',
  paymentTerms: {
    eventBase: 'BL',
    when: PaymentTermsWhen.From,
    time: 30,
    timeUnit: PaymentTermsTimeUnit.Days,
    dayType: PaymentTermsDayType.Calendar
  },
  currency: Currency.USD,
  priceUnit: PriceUnit.BBL,
  quantity: 600000,
  deliveryTerms: 'FOB',
  deliveryLocation,
  minTolerance: 1,
  maxTolerance: 1,
  invoiceQuantity: InvoiceQuantity.Load,
  generalTermsAndConditions: 'SUKO90',
  law: Law.EnglishLaw,
  commodity: 'BFOET',
  creditRequirement: CreditRequirements.OpenCredit,
  requiredDocuments: [],
  createdAt: '',
  updatedAt: '',
  tradingRole
})

export const fakeLetterOfCreditEnriched = ({
  _id = '08e9f8',
  transactionHash = '0x001',
  reference = 'LC18-MER-1',
  companyStaticId = APPLICANT_ID,
  members = [],
  tasks = [],
  trades = {},
  userId = '',
  status = ILetterOfCreditStatus.REQUESTED
} = {}): ILetterOfCreditEnriched => {
  const letter = fakeLetterOfCredit({ _id, transactionHash, reference, status })
  const role = findRole(letter, companyStaticId)
  return {
    ...letter,
    role,
    ...findParticipantCommonNames(letter, members),
    ...findLatestShipment(letter),
    actionStatus: findTaskStatusByLetterOfCreditId(tasks, letter._id),
    tasks: findTasksByLetterOfCreditId(tasks, letter._id)
  }
}

export const fakeLetterOfCreditWithLegacyVaktIdOnly = (args: any) => {
  const withVaktIdOnly = fakeLetterOfCreditWithLegacyVaktId(args)
  delete withVaktIdOnly.tradeAndCargoSnapshot.trade.sourceId
  delete withVaktIdOnly.tradeAndCargoSnapshot.cargo.sourceId

  return withVaktIdOnly
}

export const fakeLetterOfCreditWithLegacyVaktId = ({ vaktId = '1234', sourceId, ...args }) => {
  const withVaktId = fakeLetterOfCredit(args) as any
  withVaktId.tradeAndCargoSnapshot.trade.vaktId = vaktId
  withVaktId.tradeAndCargoSnapshot.cargo.vaktId = vaktId
  withVaktId.tradeAndCargoSnapshot.trade.source = TradeSource.Vakt
  if (sourceId) {
    withVaktId.tradeAndCargoSnapshot.trade.sourceId = sourceId
    withVaktId.tradeAndCargoSnapshot.cargo.sourceId = sourceId
  }
  return withVaktId
}

export const fakeTask = ({
  summary = 'fake task',
  status = TaskStatus.ToDo,
  type = LetterOfCreditTaskType.REVIEW_APPLICATION,
  actions = [] as string[],
  context = {},
  assignee = ''
} = {}): Task => {
  return {
    _id: '123',
    summary,
    taskType: type,
    status,
    counterpartyName: '(counterpartyName) remove it',
    assignee, // user ID from Keycloak or null if unassigned
    actions: mapActionsToTaskType(type),
    requiredPermission: {
      productId: 'tradeFinance',
      actionId: 'read'
    },
    context,
    outcome: undefined,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  }
}

export const fakeCargo = ({
  cargoId = '100',
  source = TradeSource.Vakt,
  sourceId = 'vakt123',
  grade = Grade.Brent as string,
  parcels = [],
  status = 'ok'
} = {}): ICargo => ({
  _id: '123',
  sourceId,
  source,
  grade,
  cargoId,
  parcels,
  createdAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  status
})

export const fakeTradeAndCargoSnapshot = ({
  source = 'VAKT',
  sourceId = 'EV111',
  trade = fakeTrade(),
  cargo = buildFakeCargo()
} = {}): ITradeAndCargoSnapshot => ({
  source,
  sourceId,
  trade,
  cargo
})

export const mapOfFakeCargo = immutable.Map<string, ICargo>({
  cargo1: buildFakeCargo,
  cargo2: buildFakeCargo
})

export const mockDate = () => {
  const RealDate = Date
  const _GLOBAL: any = global // we love typescript!

  return {
    freeze: (isoDate: string | number | Date) => {
      _GLOBAL.Date = class extends RealDate {
        constructor(...args: any) {
          super()
          return new RealDate(isoDate)
        }
      }
    },
    restore: () => {
      global.Date = RealDate
    }
  }
}

export const allLCStatusFetched = (): IStateTransition[] => {
  return [
    {
      fromState: '',
      toState: ILetterOfCreditStatus.REQUESTED,
      performer: 'Mer',
      date: '1-11-2018'
    },
    {
      fromState: ILetterOfCreditStatus.REQUESTED,
      toState: ILetterOfCreditStatus.REQUEST_REJECTED,
      performer: 'SC',
      date: '1-11-2018'
    }
  ]
}

export const fakeParcel = ({ _id = '123', id = '1234', timestamp = '2019-01-03T13:14:28.025Z' } = {}) => ({
  _id,
  id,
  laycanPeriod: {
    startDate: timestamp,
    endDate: timestamp
  },
  modeOfTransport: ModeOfTransport.Vessel,
  vesselIMO: 1,
  vesselName: 'Test',
  loadingPort: 'Test',
  dischargeArea: 'Test',
  inspector: 'Test',
  deemedBLDate: timestamp,
  quantity: 1
})

export const fakePresentation = ({
  staticId = '123',
  LCReference = '2018-BP-16',
  applicantId = 'a3d82ae6-908c-49da-95b3-ba1ebe7e5f85',
  beneficiaryId = BENEFICIARY_ID,
  documents = [],
  reference = '123',
  status = LCPresentationStatus.Draft
} = {}): ILCPresentation => ({
  staticId,
  reference,
  LCReference,
  applicantId,
  beneficiaryId,
  issuingBankId: 'a28b8dc3-8de9-4559-8ca1-272ccef52b47',
  nominatedBankId: 'a28b8dc3-8de9-4559-8ca1-272ccef52b48',
  status,
  stateHistory: [
    {
      toState: 'DRAFT',
      performer: BENEFICIARY_ID,
      date: '2019-02-19T08:22:47.686Z'
    }
  ],
  documents:
    documents.length === 0
      ? [
          {
            documentId: 'bddd54e3-4369-4627-9a9c-4f0ed186eab4',
            documentHash: 'tradedasdsadaFinance',
            status: 'DRAFT',
            documentTypeId: 'tradeFinance',
            dateProvided: SOME_DATE
          }
        ]
      : documents,
  beneficiaryComments: '',
  nominatedBankComments: '',
  issuingBankComments: ''
})

export const fakeDocument = ({
  context = {
    productId: 'tradeFinance',
    subProductId: 'lc',
    lcPresentationStaticId: '5c5441638d960da68ca87b81'
  },
  id = '1',
  name = 'AML Letter - test'
} = {}): Document => {
  const mockProduct: Product = { name: 'TRADE FINANCE', id: 'tradeFinance' }
  const mockCategory = { id: 'banking-documents', name: 'banking-documents', product: mockProduct }
  const mockType = {
    id: '2',
    name: 'type-name',
    product: mockProduct,
    category: mockCategory,
    fields: [],
    predefined: true
  }
  const COMPANY_ID = 'societegenerale'
  return {
    id,
    documentId: 'document-id',
    name,
    product: mockProduct,
    category: mockCategory,
    type: mockType,
    owner: { firstName: 'Owner', lastName: 'Owner', companyId: 'company-id' },
    hash: 'hash',
    receivedDate: new Date(SOME_DATE),
    registrationDate: new Date(SOME_DATE),
    metadata: [],
    content: undefined,
    sharedWith: [{ counterpartyId: COMPANY_ID, sharedDates: [new Date(SOME_DATE)] }],
    sharedBy: '',
    context,
    state: BottomSheetStatus.REGISTERED
  }
}

export const fakeLetterOfCreditDiff = ({
  value = 'SPLIT' as any,
  oldValue = 'OTHER' as any,
  path = '/feesPayableBy',
  op = 'replace' as any,
  type = 'ILC'
} = {}): IDiff => ({ value, oldValue, path, op, type })

export const fakeTimer = ({ time = '2019-03-30T11:10:32.523Z' } = {}): ITimer => ({
  submissionDateTime: '2019-03-20T11:10:32.523Z',
  timerData: [
    {
      id: '123',
      timerId: '123',
      time,
      status: '',
      retry: 1
    }
  ]
})
