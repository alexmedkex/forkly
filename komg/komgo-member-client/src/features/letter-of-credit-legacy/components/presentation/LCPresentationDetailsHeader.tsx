import * as React from 'react'
import { Grid, Header } from 'semantic-ui-react'
import { ILCPresentation } from '../../types/ILCPresentation'
import { capitalize } from '../../../../utils/casings'
import { findMemberName } from '../../utils/selectors'
import { displayDate } from '../../../../utils/date'
import { IMember } from '../../../members/store/types'

interface IProps {
  presentation: ILCPresentation
  members: IMember[]
}

const LCPresentationDetailsHeader: React.FC<IProps> = (props: IProps) => {
  const { presentation, members } = props
  return (
    <Grid>
      <Grid.Column floated="left" width={8}>
        <Header as="h1">{presentation.LCReference}</Header>
        <Header as="h3" style={{ marginTop: 0 }}>
          Presentation #{presentation.reference}
        </Header>
      </Grid.Column>
      <Grid.Column floated="right" width={8} textAlign="right">
        <Header as="h3">Status: {capitalize(presentation.status)}</Header>
        <p>
          Received from {findMemberName(presentation.beneficiaryId, members)}{' '}
          {presentation.stateHistory && presentation.stateHistory.length > 0
            ? `on ${displayDate(presentation.stateHistory[presentation.stateHistory.length - 1].date)}`
            : null}
        </p>
      </Grid.Column>
    </Grid>
  )
}

export default LCPresentationDetailsHeader
