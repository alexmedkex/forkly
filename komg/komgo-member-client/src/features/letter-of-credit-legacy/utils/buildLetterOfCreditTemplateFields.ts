import { ILetterOfCreditTemplate, LetterOfCreditValues } from '../constants'
import { IMember } from '../../members/store/types'
import { ITradeEnriched } from '../../trades/store/types'
import { replaceUnderscores } from '../../../utils/casings'
import { participantDetailsFromMember, findMembersByStatic } from './selectors'
import { displayPaymentTerms } from '../../../utils/date'
import { ICargo } from '@komgo/types'

export function buildLetterOfCreditTemplateFields(
  values: LetterOfCreditValues,
  members: IMember[],
  trade: ITradeEnriched,
  cargoMovements: ICargo[]
): ILetterOfCreditTemplate {
  const applicant = participantDetailsFromMember(findMembersByStatic(members, values.applicantId))
  const beneficiary = participantDetailsFromMember(findMembersByStatic(members, values.beneficiaryId))
  const issuingBank = participantDetailsFromMember(findMembersByStatic(members, values.issuingBankId))
  const beneficiaryBank = participantDetailsFromMember(findMembersByStatic(members, values.beneficiaryBankId))
  const paymentTerms = trade.paymentTerms
  const cargoMovement = cargoMovements[0] || { grade: undefined }
  return {
    ...values,
    issuingBankAddress: issuingBank.address,
    issuingBankCountry: issuingBank.country,
    beneficiaryBankAddress: beneficiaryBank.address,
    beneficiaryBankCountry: beneficiaryBank.country,
    applicant: {
      name: applicant.companyLegalName,
      locality: applicant.city
    },
    beneficiary: {
      name: beneficiary.companyLegalName,
      locality: beneficiary.city
    },
    beneficiaryBank: {
      name: beneficiaryBank.companyLegalName,
      locality: beneficiaryBank.city
    },
    issuingBank: {
      name: issuingBank.companyLegalName,
      locality: issuingBank.city
    },
    grade: cargoMovement.grade,
    cargo: {
      deliveryPeriod: {
        startDate: trade.deliveryPeriod.startDate,
        endDate: trade.deliveryPeriod.endDate
      },
      paymentTerms: displayPaymentTerms({
        eventBase: paymentTerms.eventBase,
        when: paymentTerms.when,
        time: paymentTerms.time,
        timeUnit: paymentTerms.timeUnit,
        dayType: paymentTerms.dayType
      }),
      price: trade.price,
      currency: trade.currency,
      priceUnit: trade.priceUnit,
      quantity: trade.quantity,
      deliveryTerms: trade.deliveryTerms,
      minTolerance: trade.minTolerance,
      maxTolerance: trade.maxTolerance
    },
    // Remove underscores
    applicableRules: replaceUnderscores(values.applicableRules),
    availableBy: replaceUnderscores(values.availableBy),
    // MM - Todo, hardcoded values provided until we have the proper values defined.
    // These did not break before, because we were not running that piece of the template,
    // but since we can now provide different grades, then this was now breaking
    LOIAllowed: true,
    LOI: values.LOI || ''
  }
}
