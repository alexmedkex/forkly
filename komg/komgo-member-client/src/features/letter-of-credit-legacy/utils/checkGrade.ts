import { Grade } from '@komgo/types'
import { sentenceCase } from '../../../utils/casings'

export const gradeIsAllowedValue = (grade?: string) => !!Grade[(grade ? sentenceCase(grade) : '') as Grade]
