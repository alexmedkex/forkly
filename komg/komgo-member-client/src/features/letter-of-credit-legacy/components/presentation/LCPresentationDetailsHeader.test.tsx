import * as React from 'react'
import LCPresentationDetailsHeader from './LCPresentationDetailsHeader'
import { fakeMember, fakePresentation } from '../../utils/faker'
import * as renderer from 'react-test-renderer'

describe('LCPresentationDetailsHeader', () => {
  let defaultProps

  beforeEach(() => {
    defaultProps = {
      members: [fakeMember({ staticId: '08e9f8e3-94e5-459e-8458-ab512bee6e2c', commonName: 'Applicant Name' })],
      presentation: fakePresentation({ staticId: '123', reference: '123' })
    }
  })

  it('should match snapshot', () => {
    const tree = renderer.create(<LCPresentationDetailsHeader {...defaultProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
