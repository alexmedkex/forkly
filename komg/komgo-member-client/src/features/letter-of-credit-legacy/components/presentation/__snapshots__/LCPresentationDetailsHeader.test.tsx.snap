// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LCPresentationDetailsHeader should match snapshot 1`] = `
<div
  className="ui grid"
>
  <div
    className="left floated eight wide column"
  >
    <h1
      className="ui header"
    >
      2018-BP-16
    </h1>
    <h3
      className="ui header"
      style={
        Object {
          "marginTop": 0,
        }
      }
    >
      Presentation #
      123
    </h3>
  </div>
  <div
    className="right aligned right floated eight wide column"
  >
    <h3
      className="ui header"
    >
      Status: 
      Draft
    </h3>
    <p>
      Received from 
      Applicant Name
       
      on 2019/02/19
    </p>
  </div>
</div>
`;
